xchange-stream-gemini/src/main/java/info/bitrich/xchangestream/gemini/v2/dto/GeminiOrderbook.java

```diff
@@ -0,0 +1,67 @@
+package info.bitrich.xchangestream.gemini.v2.dto;
+
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * Adapted from V1 by Max Gao on 01-09-2021
+ */
+public class GeminiOrderbook {
+  private final Map<BigDecimal, GeminiLimitOrder> asks;
+  private final Map<BigDecimal, GeminiLimitOrder> bids;
+
+  private final CurrencyPair currencyPair;
+
+  public GeminiOrderbook(CurrencyPair currencyPair) {
+    this.bids = new TreeMap<>(Comparator.reverseOrder());
+    this.asks = new TreeMap<>();
+    this.currencyPair = currencyPair;
+  }
+
+  public void createFromLevels(GeminiLimitOrder[] levels) {
+    for (GeminiLimitOrder level : levels) {
+      Map<BigDecimal, GeminiLimitOrder> orderBookSide =
+          level.getSide() == Order.OrderType.ASK ? asks : bids;
+      orderBookSide.put(level.getPrice(), level);
+    }
+  }
+
+  public void updateLevel(GeminiLimitOrder level) {
+    Map<BigDecimal, GeminiLimitOrder> orderBookSide =
+        level.getSide() == Order.OrderType.ASK ? asks : bids;
+    boolean shouldDelete = level.getAmount().compareTo(BigDecimal.ZERO) == 0;
+    // BigDecimal is immutable type (thread safe naturally),
+    // strip the trailing zeros to ensure the hashCode is same when two BigDecimal are equal but
+    // decimal scale are different, such as "1.1200" & "1.12"
+    BigDecimal price = level.getPrice().stripTrailingZeros();
+    if (shouldDelete) {
+      orderBookSide.remove(price);
+    } else {
+      orderBookSide.put(price, level);
+    }
+  }
+
+  public void updateLevels(GeminiLimitOrder[] levels) {
+    for (GeminiLimitOrder level : levels) {
+      updateLevel(level);
+    }
+  }
+
+  public Collection<GeminiLimitOrder> getAsks() {
+    return asks.values();
+  }
+
+  public Collection<GeminiLimitOrder> getBids() {
+    return bids.values();
+  }
```

Comment:
Not sure if we intend the orderbook data structure to be threadsafe here but I imagine if you call update on a side and also on a separate thread iterate over that side I suspect you will get a concurrent modification exception.

Reply:
I'll take a deeper look into this and get back to you.

Reply:
Went ahead and removed GeminiOrderBook.java and GeminiLimitOrder.java from V2 since they're not being used.

I stitched the logic from coinbasepro's stream for processing changes in Gemini's MarketDataService when I worked on this. The bids and asks are ConcurrentHashMaps which should preserve thread safety, though they're being modified by the .toOrderBook() method, which doesn't seem all that clean. All of this should run on a single thread though if I understand it correctly (the netty nio thread?).

Reply:
This should run on a single thread, but if you pass around that data structure while it's getting updated by the netty thread another thread will most likely be traversing it.